Applications/Spire/Source/KeyBindings/TagNameBox.cpp

```diff
@@ -0,0 +1,196 @@
+      if(auto i = find_name(value); i != m_list->end()) {
+        return QValidator::Acceptable;
+      }
+      return QValidator::Invalid;
+    }
+
+    QValidator::State set(const QString& value) override {
+      if(auto i = find_name(value); i != m_list->end()) {
+        return m_source->set(*i);
+      }
+      return QValidator::Invalid;
+    }
+
+    connection connect_update_signal(
+        const UpdateSignal::slot_type& slot) const override {
+      return m_update_signal.connect(slot);
+    }
+
+    auto find_name(const QString& name) const {
+      return std::find_if(m_list->begin(), m_list->end(),
+        [&] (const std::shared_ptr<AdditionalTag>& tag) {
+          return name == tag->get_name();
+        });
+    }
+
+    void on_update(const std::shared_ptr<AdditionalTag>& tag) {
+      m_update_signal(tag->get_name());
+    }
+  };
+}
+
+TagNameBox::TagNameBox(
+  std::shared_ptr<ListModel<std::shared_ptr<AdditionalTag>>> tags,
+  QWidget* parent)
+  : TagNameBox(std::move(tags),
+      std::make_shared<LocalValueModel<std::shared_ptr<AdditionalTag>>>(
+        std::make_shared<OrderFieldAdditionalTag>(Nexus::Tag(),
+          OrderFieldInfoTip::Model())), parent) {}
+
+TagNameBox::TagNameBox(
+    std::shared_ptr<ListModel<std::shared_ptr<AdditionalTag>>> tags,
+    std::shared_ptr<ValueModel<std::shared_ptr<AdditionalTag>>> current,
+    QWidget* parent)
+    : QWidget(parent),
+      m_tags(std::move(tags)),
+      m_current(std::move(current)) {
+  auto name_list_model = std::make_shared<AdditionalTagNameListModel>(m_tags);
+  m_drop_down_box = new DropDownBox(name_list_model,
+    std::make_shared<ListIndexValueModel<QString>>(name_list_model,
+      std::make_shared<AdditionalTagNameValueModel>(m_tags, m_current)),
+    ListView::default_view_builder);
+  enclose(*this, *m_drop_down_box);
+  proxy_style(*this, *m_drop_down_box);
+  setFocusProxy(m_drop_down_box);
+  m_connection = m_current->connect_update_signal(
+    std::bind_front(&TagNameBox::on_current, this));
+}
+
+const std::shared_ptr<ListModel<std::shared_ptr<AdditionalTag>>>&
+    TagNameBox::get_tags() const {
+  return m_tags;
+}
+
+const std::shared_ptr<ValueModel<std::shared_ptr<AdditionalTag>>>&
+    TagNameBox::get_current() const {
+  return m_current;
+}
+
+bool TagNameBox::is_read_only() const {
+  return m_drop_down_box->is_read_only();
+}
+
+void TagNameBox::set_read_only(bool is_read_only) {
+  m_drop_down_box->set_read_only(is_read_only);
+  if(is_read_only) {
+    create_info_tip(m_current->get());
+  } else if(m_info_tip) {
+    m_info_tip->window()->deleteLater();
+  }
+}
+
+connection TagNameBox::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_drop_down_box->connect_submit_signal([=] (const auto&) {
+    slot(m_current->get());
+  });
+}
+
+void TagNameBox::create_info_tip(
+    const std::shared_ptr<AdditionalTag>& current) {
+  if(auto tag = std::dynamic_pointer_cast<OrderFieldAdditionalTag>(current);
```

Comment:
Don't use a dynamic cast, ```TagNameBox``` there's no reason why special treatment should be given to any specific tag.

Reply:
How to fill the `OrderFieldInfoTip` up used within the `TagNameBox`? The `AdditionalTag` doesn't provide information for the `OrderFieldInfoTip` except providing the tag name.

Reply:
I don't know, the issue is if I have to spend the time to think over the design of this, then I may as well implement it myself.

Try to think through the tasks assigned to you, come up with use cases, think about possible scenarios and implementations, and then only after you have a really good understanding of the problem, proceed to implement it. When I see hacky solutions like dynamic casts and other tricks that subvert the type system it's a strong indication that the problem or the task is not well understood and wasn't properly thought through and I can't approve these solutions.

Reply:
I proposed adding a new method to `AdditionalTag` to create a tip.
 ```
 /**
   * Makes a tip displaying the information of this tag.
   * @param parent The parent of the tip.
   */
  virtual QWidget* make_info_tip(QWidget& parent) const = 0;
```

Reply:
That method looks too wide, can you narrow it down?

Reply:
I can restrict it to return the `OrderFieldInfoTip` to display the tag information.
```
/**
  * Makes an OrderFieldInfoTip displaying the information of this tag.
  * @param parent The parent of the tip.
  */
 virtual OrderFieldInfoTip* make_info_tip(QWidget& parent) const = 0;
 ```

Reply:
Why is it returning an ```OrderFieldInfoTip```? Why does it take a parent as a parameter? What is the utility of having every single derived class do this?

Reply:
Why are these virtual? Virtual functions are intended to allow for polymorphic behavior, what polymorphic behavior, or any behavior are these methods supposed to have other than returning values?

This task has gone on for too long and I don't see much progress being made.

Please close it and work on something else.